src/accelerate/hooks.py

```diff
@@ -481,14 +540,32 @@ def attach_align_device_hook_on_blocks(
             of the forward. This should only be used for classes that have submodules which are registered but not
             called directly during the forward, for instance if a `dense` linear layer is registered, but at forward,
             `dense.weight` and `dense.bias` are used in some operations instead of calling `dense` directly.
+        tied_params_map (Optional[Dict[int, Dict[torch.device, torch.Tensor]]], *optional*, defaults to `None`):
+            A map of data pointers to dictionaries of devices to already dispatched tied weights. For a given execution
+            device, this parameter is useful to reuse the first available pointer of a shared weight for all others,
+            instead of duplicating memory.
     """
+    # When dispatching the model's parameters to the devices specified in device_map, we want to avoid allocating memory several times for the
+    # tied parameters. The dictionary tied_params_map keeps track of the already allocated data for a given tied parameter (represented by its
+    # original pointer) on each devices.
+    if tied_params_map is None:
+        tied_params = find_tied_parameters(module)
+
+        tied_params_map = {}
+        for group in tied_params:
+            for param_name in group:
+                # data_ptr() is enough here, as `find_tied_parameters` finds tied params simply by comparing `param1 is param2`, so we don't need
+                # to care about views of tensors through storage_offset.
+                data_ptr = recursive_getattr(module, param_name).data_ptr()
+                tied_params_map[data_ptr] = {}
+
```

Comment:
Maybe we put this logic in `dispatch_model` since we only call it once + we already call `find_tied_parameters(module)` in `dispatch_model` just before this function.